### ISSUE TITLE
node@18.x.x version not fetch global api ?

### ISSUE THREAD

**Issue by xiaocai-rookie** [DefinitelyTyped/DefinitelyTyped#60924]
I use @types/node@18.0.0 version ,but use fetch fail,typescript to me "fetch is not defined"
Could I know , why "fetch is not defined" ?

**Comment by mahdi-farnia** [comment]
I think because it's still experimental\
Also definition for [`node:readline/promises`](https://nodejs.org/api/readline.html#promises-api) is not available

**Comment by xiaocai-rookie** [comment]
> I think because it's still experimental Also definition for [`node:readline/promises`](https://nodejs.org/api/readline.html#promises-api) is not available

Yes, node official announce too.

**Comment by shadow-light** [comment]
It was simply not added [because of time constraints](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60047#issuecomment-1147576743) so hopefully sometime soon...

**Comment by SuperchupuDev** [comment]
@SimonSchick any chance we can get `fetch` added anytime soon? it's a really important thing to have in the types

**Comment by piranna** [comment]
Is it available in any branch or development version we can make use of?

**Comment by Soldy** [comment]
LOL! Just use vanilla js that solve the problem.

**Comment by piranna** [comment]
> LOL! Just use vanilla js that solve the problem.

I would love to, but my client asked me to use Typescript :-)

**Comment by Soldy** [comment]
> > LOL! Just use vanilla js that solve the problem.
> 
> I would love to, but my client asked me to use Typescript :-)

I know your pain :+1: 
I have the same problem :)
Man you are crazy !!! :D https://github.com/piranna/coreutils.js

**Comment by SimonSchick** [comment]
The problem is still that dom libs are not modularized, if they were node typings could just reference dom specs.
As this time I'd rather not copy & paste the API definitions into node as that could create future liabilities.

**Comment by piranna** [comment]
> The problem is still that dom libs are not modularized, if they were node typings could just reference dom specs.
> As this time I'd rather not copy & paste the API definitions into node as that could create future liabilities.

Agree on that. Can you point us where it's defined so we can copy it ourselves in our projects, meanwhile it get added to Node.js definitions?

**Comment by mahdi-farnia** [comment]
I think it's better to not to use fetch Api in node. But if U need it ( for any reason ) it's seems easy to define one:

```ts
declare function fetch(...): ...;
```
1. open a new file ( js or ts ) that fetch already available
2. Command + Click or Ctrl + Click  on fetch ( goto definition )
3. Copy and paste the types to source file

If you dont wanna see it in your ts files you can just simply create types.d.ts file ( name is important ) or create a folder `@types` ( name is important ) and put it on that folder

**Note: your d.ts ( definition file ) must not be a module if it is, you have to import the fetch type in every file that you are using it.

**Comment by piranna** [comment]
> I think it's better to not to use fetch Api in node

Why not?


> simply create types.d.ts file

That's what I'm asking for.

**Comment by Soldy** [comment]
> > I think it's better to not to use fetch Api in node
> 
> Why not?
> 
> > simply create types.d.ts file
> 
> That's what I'm asking for.

Its not a good code indeed : https://github.com/nodejs/undici
But not worst than the microsoft average. https://github.com/nodejs/node/tree/v18.x/lib

**Comment by mahdi-farnia** [comment]
> > I think it's better to not to use fetch Api in node
> 
> Why not?
> 
> > simply create types.d.ts file
> 
> That's what I'm asking for.

1. It's experimental
2. Node@18 is not even the LTS version, it's the latest, not the most stable one

**Comment by piranna** [comment]
> Not a good code indeed : [nodejs/undici](https://github.com/nodejs/undici)

I would not say @mcollina code is not good, with so much performance obsesive code done...


> But not worst than the microsoft average.
> Good luck for this : [https://github.com/nodejs/node/tree/v18.x/lib](https://github.com/nodejs/node/tree/v18.x/lib?rgh-link-date=2022-07-21T07%3A55%3A15Z)

Node.js is not from Microsoft, Microsoft has only developed Typescript on top of it.

**Comment by piranna** [comment]
> 1. It's experimental

Good point, fair enough.


> 2\. Node@18 is not even the LTS version, it's the latest, not the most stable one

It's the most stable one, just only doesn't have yet the LTS denomination. I could agree that odd versions (v17, v19...) are unstable development versions and should not be used on production, but pair ones (v16, v18) is totally fine.

**Comment by Soldy** [comment]
> > Not a good code indeed : [nodejs/undici](https://github.com/nodejs/undici)
> 
> I would not say @mcollina code is not good, with so much performance obsesive code done...

[undici](https://github.com/nodejs/undici) fetch is a heavy weight curl. I do not like it. Sorry.
I learned a lot from @[mcollina](https://github.com/mcollina) code, but most of my node code are much older, and do a better performance.  

> 
> > But not worst than the microsoft average.
> > Good luck for this : [https://github.com/nodejs/node/tree/v18.x/lib](https://github.com/nodejs/node/tree/v18.x/lib?rgh-link-date=2022-07-21T07%3A55%3A15Z)
> 
> Node.js is not from Microsoft, Microsoft has only developed Typescript on top of it.

I never said the microsoft built the nodejs (that was Ryan Dahl, and the openjs foundation, @[BethGriggs](https://github.com/BethGriggs) ... etc ). They just damaged with typescript :P 
The problem I worked with nodejs because the lightweight env. I built my drones with javascript first after rewrite in C and ADA. 
I have limited memory and limited storage. AND NO THIRD PARTY DEPENDENCIES :D 
Now I have to go back to python and rust. 
So I rewrite everything in opensource to support my ex client.

**Comment by rjoonas** [comment]
Yeah, node 18 will be LTS in two months (2022-10-25).

> The problem is still that dom libs are not modularized, if they were node typings could just reference dom specs. As this time I'd rather not copy & paste the API definitions into node as that could create future liabilities.

@SimonSchick, so are we expecting lib.dom to be modularized before we can add node typings for fetch? Sounds like something that might take a while.

I wonder if in the meanwhile we could add fetch to types/node by copy-pasting fetch, Request, Response, Headers etc. definitions as a standalone module that would be easy to delete when it's possible to just reference the respective DOM definitions.

**Comment by piranna** [comment]
> I wonder if in the meanwhile we could add fetch to types/node by copy-pasting fetch, Request, Response, Headers etc. definitions as a standalone module that would be easy to delete when it's possible to just reference the respective DOM definitions.

Agree, and add a `TODO:` comment about that fact, or maybe also store them in an independent types file and reference it from the `lib.dom` types... starting with the modularization ourselves :-)

**Comment by piranna** [comment]
Or just import the `lib.dom` types, and re-export the ones we are interested about.

**Comment by brielov** [comment]
Any news on this? I'd prefer not to have my global namespace polluted with `lib.dom` stuff.

**Comment by ImRodry** [comment]
> The problem is still that dom libs are not modularized, if they were node typings could just reference dom specs. As this time I'd rather not copy & paste the API definitions into node as that could create future liabilities.

Is there any chance you could install undici as a dep on `@types/node` and re-export the fetch types from it @SimonSchick?
Also does anyone know if the fetch API will be considered stable by the time Node v18 hits LTS?

**Comment by mcollina** [comment]
We are not planning to have a stable fetch for when v18 hits LTS.

**Comment by llwwbb** [comment]
> We are not planning to have a stable fetch for when v18 hits LTS.

I don't quite understand whether node18 is lts or not, and whether fetch is experimental or not, what does it have to do with whether there is a definition of fetch in @types/node? Has anyone stipulated that all the apis defined in @types are stable? I do not think so. @types is just a type definition. Anything that can be called or accessed in js should have a corresponding type definition, regardless of whether it is experimental or not.

**Comment by piranna** [comment]
I'm not so much concerned about being stable or not... Just only, i think if it's available without flags, It should be available for the same version. It's said, @types/node 18.0.0 should have all the types of Node.js 18, i can be happy of not having available a newer version or being It delayed until it's complete and on pair, and need to use and stick with @types/node 17.x.x, than install 18.x.x thinking It has all the same types and have some of them missing. At least, if It will provide only non-experimental ones, notify It in the readme.

**Comment by HoldYourWaffle** [comment]
> Or just import the `lib.dom` types, and re-export the ones we are interested about.

This wouldn't work because merely touching `lib.dom.d.ts` pollutes the global scope.

However, as a temporary workaround you could add a declaration file to your project that re-exports the types from [`@types/node-fetch`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node-fetch/index.d.ts):

```ts
import _fetch from 'node-fetch'

declare global {
    declare var fetch: typeof _fetch;
}
```

`node-fetch`'types [aren't 100% accurate](https://github.com/node-fetch/node-fetch/blob/2.x/LIMITS.md) for native `fetch` of course, but depending on your preferences it could be a good workaround compared to having no types at all.

**Comment by piranna** [comment]
I like the workaround, thanks

**Comment by thw0rted** [comment]
I just noticed that this issue does not have a link to https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1207 . As @HoldYourWaffle pointed out, the current lib-dom does not export types, it only makes global declarations.  It would be great if those types were available for "clean" import, but it doesn't seem to be a priority yet.  Maybe an upvote for that issue would help to solve this one?

**Comment by piranna** [comment]
> I just noticed that this issue does not have a link to [microsoft/TypeScript-DOM-lib-generator#1207](https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1207) . As @HoldYourWaffle pointed out, the current lib-dom does not export types, it only makes global declarations. It would be great if those types were available for "clean" import, but it doesn't seem to be a priority yet. Maybe an upvote for that issue would help to solve this one?

Upvoted and commented at https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1207#issuecomment-1236992985

**Comment by Qix-** [comment]
What does experimental status, DOM modularity, or LTS have **anything** to do with whether or not types are included here? I don't see the point in arguing _any_ of that - the point is, it exists in Node 18 without a flag. There should be types for it. No, sorry, DOM is not acceptable because Node does not use the DOM. WhatWG Fetch is a standard pulled out of the browsers and re-implemented in Node.js becuase it is useful. The fact it originated in browsers is completely irrelevant.

_Why is this a discussion?_ This is a hole in the types, plain and simple.

**Comment by thw0rted** [comment]
The reason I brought up DOM modularity is basically https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59905 .  Lots of people wind up with the `@types/node` included in their frontend project, or `lib: ['dom']` in their backend project, for a wide variety of reasons.  If `@types/node` ships a global `fetch`, lots of users will suddenly find that they are merging those types with the ones built into lib-dom.  I'm not sure what the result will be, or even how to test it, but this impedance-mismatch has caused problems before and I suspect they're about to get a lot worse.

**Comment by thw0rted** [comment]
Also while I'm thinking of it, re @HoldYourWaffle 's comment about adding your own types for this, I wanted to suggest

```ts
declare global {
  var fetch: typeof import("node-fetch").default;
}
```

I believe this makes it clear to the typechecker that you're 100% not trying to import `node-fetch` at runtime, and only care about the types.

Of course I do still think we need a path to correct types directly in `@types/node`, I'm just not sure what that's going to look like.

**Comment by Qix-** [comment]
Node.js uses `undici` under the hood for `fetch`. Not sure if it's exposed directly or if it's wrapped, but it comes with Typescript types that could be consumed here somehow.

**Comment by piranna** [comment]
> Node.js uses `undici` under the hood for `fetch`. Not sure if it's exposed directly or if it's wrapped, but it comes with Typescript types that could be consumed here somehow.

Definitely a good idea, can somebody confirm?

**Comment by meyfa** [comment]
> Node.js uses undici under the hood for fetch. Not sure if it's exposed directly or if it's wrapped, but it comes with Typescript types that could be consumed here somehow.

This is true. As a workaround, I recommend instead of node-fetch, adding undici as a dev dependency and using the following adapted declaration:

```ts
declare global {
  var fetch: typeof import('undici').fetch
}
```

**Comment by Qix-** [comment]
@piranna https://github.com/nodejs/node/blob/12fb94c78c7c0e5944db425f40c4bf3ed61b4f76/lib/internal/process/pre_execution.js#L204-L217

**Comment by piranna** [comment]
> @piranna [nodejs/node@`12fb94c`/lib/internal/process/pre_execution.js#L204-L219](https://github.com/nodejs/node/blob/12fb94c78c7c0e5944db425f40c4bf3ed61b4f76/lib/internal/process/pre_execution.js#L204-L219)
> 
> Specifically, `undici`s `module.exports.fetch` is used as Node's global `fetch` - @meyfa has done it correctly above.

I mean, confirm if it can be directly referenced, without needing to install `undici` as a devDependency, so also `@types/node` can reference it too and we can fix this issue.

**Comment by thw0rted** [comment]
Oh you beat me to it, I just left a comment on my linked PR to the effect that `@types/node` should be grabbing the types from `undici`.  (I think that *might* mean it would have to be a devDependency of `@types/node`, though?)

**Comment by thw0rted** [comment]
Yes you're both right, I just checked.  `@types/node-fetch` has a `package.json` of

```json
{
    "private": true,
    "dependencies": {
        "form-data": "^3.0.0"
    }
}
```

When installed from NPM, the package.json for `@types/node-fetch` includes

```json
    "dependencies": {
        "@types/node": "*",
        "form-data": "^3.0.0"
    },
```

This results in `form-data` being installed under `./node_modules/@types/node-fetch/node_modules/form-data`
```

**Comment by Qix-** [comment]
Not sure what the deployment process of DT packages is, but could we include `undici` as a `devDependency`, run `tsc` on it to get the type exports, then bake them into the published bundle in a `prepare` script somehow?

**Comment by thw0rted** [comment]
Yeah I think the "somehow" is the hard part -- AFAICT there is not a process within DT for extracting types from a 3rd party lib but ignoring the runtime code.  They have an allow-list for packages that a type-package may list as a dep, then they emit the dep as I showed above.  I was musing about this over on the TS Discord just now but talked myself into thinking that our use-case here is probably uncommon.  (How often do you want types for library A, which needs types from library B, but library A doesn't actually use library B at runtime?)

**Comment by buffet-time** [comment]
Just curious, is the reason there still isn't a type because no one has done it? or is there arguments about it. If its just no one putting in the work. I will do it

**Comment by HoldYourWaffle** [comment]
@buffet-time The main issue is that the types already exist in `lib.dom` and it's undesirable to duplicate them (https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924#issuecomment-1191074115).

**Comment by buffet-time** [comment]
> @buffet-time The main issue is that the types already exist in `lib.dom` and it's undesirable to duplicate them ([#60924 (comment)](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924#issuecomment-1191074115)).

So what would be the course of action I could take to get this going? I myself really would like to contribute so that more devs can have a better Type experience with 18+. I have the bandwidth in my freetime and am very interested so if someone can point me in the direction of what needs to be done to properly get this done.

**Comment by SuperchupuDev** [comment]
> > @buffet-time The main issue is that the types already exist in `lib.dom` and it's undesirable to duplicate them ([#60924 (comment)](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924#issuecomment-1191074115)).
> 
> So what would be the course of action I could take to get this going? I myself really would like to contribute so that more devs can have a better Type experience with 18+. I have the bandwidth in my freetime and am very interested so if someone can point me in the direction of what needs to be done to properly get this done.

I think https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924#issuecomment-1246622957 could be a really good way to do it, the hard way would be trying to make it work while having `undici` as a `devDependency` as mentioned above

**Comment by joehillen** [comment]
> We are not planning to have a stable fetch for when v18 hits LTS.

@mcollina  Can you please explain the reason for this? If it's in v18, why delay it? It does not make sense to me (or anyone else in this thread).

**Comment by SimonSchick** [comment]
@sandersn kinda out of the loop, what's the progress on modularizing the dom typings to avoid duplication, I think that's the major blocking factor here.

**Comment by thw0rted** [comment]
I think it just isn't a priority for the team. I linked to the issue above, but there hasn't been any activity for a while.

Maybe if somebody forked the DOM lib generator and made it create modular types, they'd be willing to accept a PR? I haven't looked at how much work this would be.

**Comment by Qix-** [comment]
**Please do not use the DOM modules in Node.js They are not the same.** Please use the `undici` types instead. **They are not the same.** For example, `undici` introduces the `dispatcher` option, which is not present in WHATWG's `fetch()` (i.e. the DOM).

**Comment by Qix-** [comment]
This has nothing to do with Typescript and everything to do with Node's types. Node.js's `undici` implements the _minimum_ interface required by the WHATWG standard, plus a few extra things. Those extra things _should_ be represented in the type system. Using `lib:['dom']` is a hack to work around the lack of these proper types. It is by no means _correct_.

**Comment by thw0rted** [comment]
I must have miscommunicated, I'm agreeing with you.  I'm not saying that backend code should include `lib:['dom']`, in fact the whole point of my linked PR is to *remove* a reference to lib-dom from Node types.  What I'm saying is, people with *frontend* projects (who obviously should include lib-dom) are also getting Node types, because they have not taken steps to explicitly exclude them.  These frontend projects get the Node version of various globals, like setTImeout and Blob, and now when we add `fetch` they will get the Node version of fetch as well.

That's the problem I'm worried about, more of these DOM-like-but-not-quite types are going to bleed through to users who want a real DOM environment.  It's not an argument against adding correct (not-actually-DOM) types to Node, it's more of a warning that when Node gets more overlapping globals, a lot of frontend users are going to suffer collateral damage.

**Comment by piranna** [comment]
> What I'm saying is, people with _frontend_ projects (who obviously should include lib-dom) are also getting Node types, because they have not taken steps to explicitly exclude them.

They have not done their work properly, and they introduced a bug on their code. Not something I would care myself, beyond adding a comment in the changelog and publish as a major version.

**Comment by thw0rted** [comment]
> They have not done their work properly, and they introduced a bug on their code

Do you mean that every frontend project you've made has an explicit `typeRoots` or `types` in the tsconfig, and anyone who doesn't do that has "introduced a bug"?  If so, could you show me where in the handbook or other official docs this is a recommended best practice?

I would contend that, for the average user, this is 1) a lot of extra work to fix something that "should" work out of the box, and 2) esoteric enough that they probably don't even realize anything is wrong in the first place.  The default behavior -- add a few simple deps, `npm install`, `npx tsc --init` -- should produce a working project without obvious bugs.  Of note, the default tsconfig does enable `strict` checks, but does *not* set an explicit `types` / `typeRoots`.

It would be one thing for some esoteric `@types` package introduced a global type that conflicted directly with the DOM, but this is `@types/node`, with over 76M weekly downloads, and 26k packages depend on it.   You can't just go around breaking 76M users' builds with a blithe "comment in the changelog", and you can't change the major version because DT major versions must match the major version of the package they're describing -- a [known limitation of semver on DT](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/25677) for a long time now.

I'm not saying that the subject change here (`fetch` types) will be the straw that breaks the camel's back, but I want to make sure everyone advocating for new `@types/node` globals is aware of the scope of the impact of such changes.

**Comment by Qix-** [comment]
I'm really not sure what the issue is you're describing @thw0rted - perhaps you can elaborate a bit.

Under which circumstances would a frontend project have `@types/node` as a dependency that could not be otherwise configured more properly?

**Comment by thw0rted** [comment]
Of course frontend packages should not *declare* a dependency on `@types/node`, but I don't know that I've ever seen a frontend project that doesn't wind up with `@types/node` installed.

Remember how I said 26k packages depend on it?  Look at [the list here](https://www.npmjs.com/browse/depended/@types/node), and try to imagine making a non-trivial frontend project with *none* of those packages, either as a direct dependency, or transitive.  On the first page alone, you've got `@types/webpack`, `@types/glob` (used by almost every CLI tool), various `jest` packages and `cypress` (also `karma` and `@types/karma`, incidentally).  Page two has `@types/rimraf` and `@types/mkdirp` (hope you're writing your `scripts` cross-platform!), `@types/puppeteer` (very common for driving e2e tests), `@types/superagent`, the list just goes on and on.

If you depend on any of those, congratulations, you just polluted your global namespace with Node types, assuming you're using the default tsconfig.

**Comment by piranna** [comment]
> On the first page alone, you've got `@types/webpack`, `@types/glob` (used by almost every CLI tool), various `jest` packages and `cypress` (also `karma` and `@types/karma`, incidentally). Page two has `@types/rimraf` and `@types/mkdirp` (hope you're writing your `scripts` cross-platform!), `@types/puppeteer` (very common for driving e2e tests), `@types/superagent`, the list just goes on and on.

All of these are server side and/or cli tools, I still don't see the need to use `@types/node` in the client side code...

**Comment by thw0rted** [comment]
Right, the problem is not that you use a CLI tool *in* client side code, you use it to *build* client side code, and the default configuration for typescript is to include all types that are installed in `node_modules/@types`.  So, you install a `devDependency` for, say, `webpack`, realize you also need `@types/webpack` -- now you just added `node_modules/@types/node`, which means (unless you take the extraordinary step of manually configuring `types` / `typeRoots`!) you have introduced Node types in your frontend code.

**Comment by Qix-** [comment]
Does `tsconfig` not honor `package.json` in terms of what is listed as a types-providing dependency?

At any rate, this problem existed before this, and I don't see how it's relevant in this issue. This would be the case for any common globals. It is indeed something for the TS folks to handle, IMO - not a concern for this issue.

**Comment by thw0rted** [comment]
No, `tsconfig` does not differentiate between `dependencies` vs `devDependencies` vs transitive dependencies.  It just has a `typeRoots` property which defaults to `node_modules/@types`; any packages that are installed there are used by default.

Yes, the problem existed before this, but it's getting worse and this issue is very much a part of the problem.  It's not practical to say "the TS folks will handle it" and ignore the issue.  As I said above, in this specific case, having both `fetch` types merged might not cause end-user conflicts, but it's definitely going to bite us eventually if we keep going down this path.

**Comment by piranna** [comment]
@thw0rted now I see the issue, and yes, server-side types are set as global too for client side code, and that's bad... :-/ But as @Qix- said, it's a bug on Typescript side then, it should not load at least transitive dependencies, and maybe also types of devDependencies, since they are types of code that's not the actual main code base.

**Comment by thw0rted** [comment]
I mean, it would be real neat if it looked at package.json to determine which types to include, but it doesn't, and never has. I don't expect the documented behavior to change any time soon, I'm just describing the situation we find ourselves in now.

**Comment by piranna** [comment]
> I mean, it would be real neat if it looked at package.json to determine which types to include

Totally agree here.

**Comment by meyfa** [comment]
I would argue that any discussion about type inclusion via `dependencies` vs `devDependencies`, though it has merit, is off-topic for this particular issue. Still, for future reference, I'd like to quickly note that types provided by `devDependencies` may very well be needed during building, unit testing, etc. even if they are never used on the production side.

**Comment by piranna** [comment]
> I'd like to quickly note that types provided by `devDependencies` may very well be needed during building, unit testing, etc. even if they are never used on the production side.

For config files, maybe?

**Comment by ZatsuneNoMokou** [comment]
Current state ?

**Comment by Qix-** [comment]
Issue is still open, clearly.

**Comment by abrkn** [comment]
Any update on this? Node v18.12.0 is now LTS.

**Comment by thw0rted** [comment]
My big no-DOM PR (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59905) and several related updates have been merged.  There's now a pattern for conditional types to only expose global types when lib-dom is not included, and test that the conditional types are correctly applied in both environments.

I wouldn't mind adding these but am going to be crazy busy for the next 2-ish weeks.  Anyone who wants to tackle it should have a look at the changes in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62782 -- separate tests should be added for DOM vs not-DOM, following the example in `test/dom-events.ts` (and `ts4.8/test/dom-events.ts`).

**Comment by WildXBird** [comment]
simply
 
![image](https://user-images.githubusercontent.com/32768476/198872205-27c0b87e-c82a-488f-b750-ec1fb9e388eb.png)

**Comment by mkp7** [comment]
for now using `"lib": ["es2021", "DOM"]` in tsconf

**Comment by Qix-** [comment]
Please read the thread before commenting here, your "fixes" are not correct and do not solve the problem and have already been discussed at length above. Every time you comment, you chip away at the time of a non-trivial amount of people.

**Comment by cjnoname** [comment]
any progress on it?

**Comment by Qix-** [comment]
The ticket is still open, so no. No progress has been made. You're welcome to file a PR. Commenting "any progress" doesn't help, there are no backchannels discussions happening. If it's not updated here, it's not updated.

**Comment by tobq** [comment]
Node 18 is now the LTS - with non-experimental native fetch support

**Comment by Qix-** [comment]
There's no calendar. All changes here are contributor-supplied. The question of how exactly to go about it is still up in the air.

**Comment by dpraul** [comment]
Here's an updated version of the workaround in https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924#issuecomment-1246622957 for Node 18.12.1 LTS, since it [exports more objects from `undici`](https://github.com/nodejs/node/blob/v18.12.1/lib/internal/process/pre_execution.js#L219-L239). FYI Node 18.12.1 appears to be using undici 5.11.0 internally, if you'd like to guarantee the types match exactly.

```ts
declare global {
  export {
    fetch,
    FormData,
    Headers,
    Request,
    Response,
  } from 'undici';
}
```

**Comment by Cuppachino** [comment]
@dpraul's workaround got rid of the red squiggles, but the types were still `any`.  Also was giving me this error:

> Exports and export assignments are not permitted in module augmentations.ts(2666)

This is what works for me:

```ts
export {}

declare global {
  export const {
    fetch,
    FormData,
    Headers,
    Request,
    Response
  }: typeof import('undici')
}
```

**Comment by cherryblossom000** [comment]
I had to add an `export {}` to @dpraul’s solution:

```ts
declare global {
	export {fetch, FormData, Headers, Request, Response} from 'undici'
}

export {}
```

**Comment by cherryblossom000** [comment]
I was using some of the types from undici, so here’s what worked for me:

```ts
declare global {
	export const {
		fetch,
		FormData,
		Headers,
		Request,
		Response
	}: typeof import('undici')
	export type {FormData, Headers, Request, Response} from 'undici'
}

export {}
```

**Comment by thernstig** [comment]
Just for reference, https://nodejs.org/en/blog/release/v18.13.0/ removed the warning for experimental for the builtin `fetch` API. It is also considered ready for general use, see https://github.com/nodejs/node/pull/45684#issuecomment-1333742929

**Comment by piranna** [comment]
> Just for reference, [nodejs.org/en/blog/release/v18.13.0](https://nodejs.org/en/blog/release/v18.13.0/) removed the warning for experimental for the builtin `fetch` API. It is also considered ready for general use, see [nodejs/node#45684 (comment)](https://github.com/nodejs/node/pull/45684#issuecomment-1333742929)

Does it support full duplex?

**Comment by pulli23** [comment]
what is the status of this bug? is there any chancethis will be added (ever)?

**Comment by Qix-** [comment]
The current status of the bug is right there at the top.

![image](https://user-images.githubusercontent.com/885648/216092656-afdced72-bc15-470c-978e-2e87948d4768.png)

**Comment by tobq** [comment]
😂 

> The current status of the bug is right there at the top.
> 
> ![image](https://user-images.githubusercontent.com/885648/216092656-afdced72-bc15-470c-978e-2e87948d4768.png)

**Comment by pulli23** [comment]
Talking more about the status of the discussion, is there consensus on how/if this is implemented? Since apparently the problem is that it would cause breaking points for people who both include `dom` library and node typings.

**Comment by thw0rted** [comment]
I believe the least damaging thing we could come up with at the moment would be to implement something like https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62782, specifically a conditional type akin to `typeof globalThis extends { onmessage: any, fetch: any } ? {} : typeof import('undici').fetch`.  I haven't had the spare cycles to do it myself yet, but I did add something like that to one of my own projects recently as a workaround.

I would be hesitant to implement that in `@types/node` today just because it would add a dependency on `undici` which would increase the installed size of the type package by about 50%.  There is an open question about how to have DT import a type from another package without also making users install the whole package to use the types; I'm not sure whether this is being tracked as its own issue or not.

**Comment by JasonMan34** [comment]
@Qix- 
> The current status of the bug is right there at the top.
> 
> ![image](https://user-images.githubusercontent.com/885648/216092656-afdced72-bc15-470c-978e-2e87948d4768.png)

What a rude, tongue-in-cheek answer to an issue that should have been resolved 9 months ago

**Comment by thw0rted** [comment]
OK so the screenshot was a bit of a smart-ass reply, but, how "should" the issue have been resolved?  First, I don't think `fetch` was out from behind the experimental flag 9 months ago.  Second, Node native `fetch` is slightly incompatible with DOM fetch, and will cause conflicts if typed correctly, as described in great detail in the comment thread above.  Third, who exactly was supposed to have resolved this?  These are community contributed types; even if we had a clear way forward, whose job is it supposed to be to implement them?

**Comment by ehartford** [comment]
If Node's official fetch api can't be included in @types/node, could we at least have an "official" workaround?

**Comment by thw0rted** [comment]
Node does not "officially" support TypeScript at all.  The types published here in DT are all community-contributed.  No one has contributed an update to add these types, so in the interim your best bet is to use a workaround like the ones a few pages upthread.  There are 3 good examples to follow in comments from mid-December.

**Comment by ehartford** [comment]
I never claimed that Node supports TypeScript.
I claimed that Node supports its fetch API.  

@types/node is supposed to provide types for all of Node's APIs.  fetch is one of those APIs, as of 18.x.

**Comment by KantiKuijk** [comment]
@types/node _aims_ to provide all the types, I do not find any supposed guarantee that says it _is_ complete. There are only types for what the community has contributed typing for (I'm sorry if I'm stating the obvious here). But since that same community isn't able to agree on a single best way to implement typing of the fetch API, the best there currently is are the workarounds mentioned previously (offered by members of the community). Whether those are "official" or not is a question of what "official" means in a community-driven context.

**Comment by tilgovi** [comment]
If there is a satisfactory typing that doesn't cause horrible conflicts for isomorphic packages, would we all be satisfied with a manual process of taking snapshots of undici types here? Could we put a script to do so in the node types directory, even if it's not run by any continuous integration, but something community maintainers do when we need to update the types?

**Comment by jordanebelanger** [comment]
How is this not fixed yet lmao

**Comment by ehartford** [comment]
> How is this not fixed yet lmao 

To be fair it wasn't highly publicized.

It makes sense that the community is slow to embrace it.  Change is hard.

**Comment by jordanebelanger** [comment]
Ya sorry for being snarky haha, but seriously fetch is a huge addition and it's node LTS with no experimental flag now, surely a lot of people are facing this issue?

**Comment by Qix-** [comment]
Surely you can read the rest of this thread, remembering that OSS is the culmination of free time for the most part, donated to these projects without much, if any, compensation.

Surely you can see that there has been no resolution as to how to pull in the third-party library's types, right? Until that's been decided upon, and a PR submitted, and approved, and released, then yeah. Things remain as they are.

Your snarky comments do absolutely _nothing_ to help improve the situation. It's not like there's a singular lazy nerd in a basement somewhere waiting for someone to remind them this issue exists in order to "get 'er done" or something.

Be better.

**Comment by revmischa** [comment]
> How is this not fixed yet lmao

If you are unsatisfied, you may open a pull request fixing it as this is an open-source project. If that is unsatisfactory you may be entitled to a full refund.

**Comment by jamiehaywood** [comment]
is there a thread where we can start talking about implementation of these types, rather than having this pointless back and forth?

**Comment by Qix-** [comment]
Yes, right here. Scroll up and read the discussion. 🙄

**Comment by emcpow2** [comment]
perpaps, `fetch` can be hidden in node prefix import?

sample

```typescript
import type { fetch } from "node:fetch";
```

That way it won't create a conflict with browser `fetch` global variable.

**Comment by mindlid** [comment]
declare var fetch: typeof import('undici').fetch;

**Comment by mindlid** [comment]
```ts
// fetch.d.ts
declare var fetch: typeof import('undici').fetch;
```

**Comment by Simonpedro** [comment]
> 

In case it helps, I also needed to define some other types

```ts
// fetch.d.ts
declare var fetch: typeof import("undici").fetch
declare type RequestInfo = import("undici").RequestInfo
declare type RequestInit = import("undici").RequestInit
declare type Request = import("undici").Request
declare type Response = import("undici").Response
declare type Headers = import("undici").RequestHeaders
```

**Comment by Simonpedro** [comment]
> > > 
> > 
> > 
> > In case it helps, I also needed to define some other types
> > ```ts
> > // fetch.d.ts
> > import type * as unidici from "undici";
> > 
> > declare global {
> >   var fetch: typeof unidici.fetch
> >   class Request extends unidici.Request { }
> >   class Response extends unidici.Response { }
> > }
> > ```
> 
> Thank you! Definitely helpful. I'm not sure if extending the Request/Response classes are necessary though. You can just destructure multiple functions and classes:
> 
> ```ts
> // fetch.d.ts
> import type * as undici from "undici";
> 
> declare global {
>   // Re-export undici fetch function and various classes to global scope.
>   // These are classes and functions expected to be at global scope according to
>   // Node.js v18 API documentation.
>   // See: https://nodejs.org/dist/latest-v18.x/docs/api/globals.html
>   export const { 
>     FormData,
>     Headers,
>     Request,
>     Response,
>     fetch,
>   }: typeof undici;
> }
> ```

Thanks @AndrewJo. Something like that was my first try, but it raises the following kind of error for me:
```
src/lib/response-serializer.ts:1:51 - error TS2749: 'Response' refers to a value, but is being used as a type here. Did you mean 'typeof Response'?

1 export async function serializeResponse(response: Response): Promise<string> {
``` 

So for now I'll stick to the "extends" approach (it doesn't look the best, but works 😅 )

**Comment by AndrewJo** [comment]
@Simonpedro oops, I think the type definitions also need to be re-exported in addition to the function and class definitions. Can you try this?

```typescript
// fetch.d.ts
import {
  type FormData as FormDataType,
  type Headers as HeadersType,
  type Request as RequestType,
  type Response as ResponseType,
} from "undici";

declare global {
  // Re-export undici fetch function and various classes to global scope.
  // These are classes and functions expected to be at global scope according to Node.js v18 API
  // documentation.
  // See: https://nodejs.org/dist/latest-v18.x/docs/api/globals.html
  export const { FormData, Headers, Request, Response, fetch }: typeof import("undici");

  type FormData = FormDataType;
  type Headers = HeadersType;
  type Request = RequestType;
  type Response = ResponseType;
}
```

**Comment by Simonpedro** [comment]
@AndrewJo It worked like charm! Thank you!

**Comment by JavaScriptBach** [comment]
@AndrewJo this snippet works for me; thank you very much! All the other ones in this thread I've tried were wrong in some way due to not properly exporting types, and I didn't find the TypeScript docs to be very instructive in writing declaration files either.

**Comment by AndrewJo** [comment]
For those that are needing to stub global fetch using libraries like Sinon or Jest, you will also need to modify `globalThis` object.

Few notes/caveats:

- Works for TypeScript 4.3 or newer

- File must be renamed from `*.d.ts` extension to `*.ts` or you will get `ts(1036)` error: 
  > Statements are not allowed in ambient contexts.

- `const` must be renamed to `var` or you will get `ts(2339)` error: 
  > Property '\<propertyName\>' does not exist on type 'typeof globalThis'.

```typescript
// fetch.ts
import {
  type FormData as FormDataType,
  type Headers as HeadersType,
  type Request as RequestType,
  type Response as ResponseType,
} from "undici";

declare global {
  // Re-export undici fetch function and various classes to global scope.
  // These are classes and functions expected to be at global scope according to Node.js v18 API
  // documentation.
  // See: https://nodejs.org/dist/latest-v18.x/docs/api/globals.html
  // eslint-disable-next-line no-var
  export var { FormData, Headers, Request, Response, fetch }: typeof import("undici");

  type FormData = FormDataType;
  type Headers = HeadersType;
  type Request = RequestType;
  type Response = ResponseType;
}

// NOTE: the import in the global block above needs to be a var for this to work properly.
globalThis.fetch = fetch;
globalThis.FormData = FormData;
globalThis.Headers = Headers;
globalThis.Request = Request;
globalThis.Response = Response;
```

Then in your test file, TypeScript will no longer complain when doing something like this:

```typescript
it("should fetch data", async function () {
  const fetchStub = sinon.stub(global, "fetch");

  // ...
});
```

**Comment by AndrewJo** [comment]
For those that are needing to stub global fetch using libraries like Sinon or Jest, you will also need to modify `globalThis` object.

Few notes/caveats:

- Works for TypeScript 4.3 or newer

- File must be renamed from `*.d.ts` extension to `*.ts` or you will get `ts(1036)` error: 
  > Statements are not allowed in ambient contexts.

- `const` must be renamed to `var` or you will get `ts(2339)` error: 
  > Property '\<propertyName\>' does not exist on type 'typeof globalThis'.

```typescript
// fetch.ts
import {
  type FormData as FormDataType,
  type Headers as HeadersType,
  type Request as RequestType,
  type Response as ResponseType,
} from "undici";

declare global {
  // Re-export undici fetch function and various classes to global scope.
  // These are classes and functions expected to be at global scope according to Node.js v18 API
  // documentation.
  // See: https://nodejs.org/dist/latest-v18.x/docs/api/globals.html
  // eslint-disable-next-line no-var
  export var { FormData, Headers, Request, Response, fetch }: typeof import("undici");

  type FormData = FormDataType;
  type Headers = HeadersType;
  type Request = RequestType;
  type Response = ResponseType;
}

// NOTE: the import in the global block above needs to be a var for this to work properly.
globalThis.fetch = fetch;
globalThis.FormData = FormData;
globalThis.Headers = Headers;
globalThis.Request = Request;
globalThis.Response = Response;
```

Then in your test file, TypeScript will no longer complain when doing something like this:

```typescript
it("should fetch data", async function () {
  const fetchStub = sinon.stub(global, "fetch");

  // ...
});
```

**Comment by lisonge** [comment]
emmm, Even if I don't use `node@18`, why `TS Error Cannot find name 'fetch'.` in `@types/node@20` and `node@20.3.0` ?

**Comment by thw0rted** [comment]
I get that this is a long thread and there's been a lot of back and forth, so it can be hard to understand the current state of the problem.

The short version is that no one has implemented the types for `fetch` in this package yet.  There are some ideas about how best to handle it but I don't know if there is actually a consensus among maintainers of this package.  If you'd like to help with that, DT has good guidelines for getting started as a contributor, and I'm sure several people on this thread would be eager to review a MR.

If you look at the most recent couple of pages of comments, there are a number of suggestions for working around it, typically by importing types from the `undici` package (which is actually used to implement `fetch` in core Node now) and re-declaring them in the global scope.  If you review the comments and still have trouble implementing that workaround you could ask for help here.

**Comment by pauldraper** [comment]
> If you look at the most recent couple of pages of comments, there are a number of suggestions for working around it, typically by importing types from the undici package (which is actually used to implement fetch in core Node now) and re-declaring them in the global scope. 

Works great.

If only there were a library that already did this :/

**Comment by paul23-git** [comment]
Weirdly as undici is now considered part of node(?) when I use `npm i undici` nothing is being installed. So creating a `fetch.ts` as explained above isn't working either now.

**Comment by VityaSchel** [comment]
> > How is this not fixed yet lmao
> 
> If you are unsatisfied, you may open a pull request fixing it as this is an open-source project. If that is unsatisfactory you may be entitled to a full refund.

Группа токсиков

**Comment by Andrewmat** [comment]
> I would be hesitant to implement that in `@types/node` today just because it would add a dependency on `undici` which would increase the installed size of the type package by about 50%. There is an open question about how to have DT import a type from another package without also making users install the whole package to use the types; I'm not sure whether this is being tracked as its own issue or not.

I think this is where the issue has stopped, so I'll continue from here.

Right now, the most recommended option for people is to install the `undici` package themselves and copy paste a snippet. This is not much different from the actual package being a dependency of this DT.

Although there is a huge increase in download size, I don't think this is a actual issue. As you have already commented, 800kb would not be a problem if this is not included in web bundles.

The conflict between Node and DOM is what worries me. Even if `@types/node` uses the hypothetical modular DOM definition, I think we are sidestepping the main issue, in that DOM projects cannot declare dependencies that adds `@types/node` to node_modules. Without considering this, I would argue that adding `undici` as dependency would be the best solution.

Some months ago I had an issue when updating webpack 4 to 5 where the Node global `Buffer` was declared in the `@types/node` but webpack [removed the polyfills from the build](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#automatic-nodejs-polyfills-removed). I fixed it using eslint, and opened a suggestion in `create-react-app` to help developers to avoid doing the same mistake (https://github.com/facebook/create-react-app/issues/13032).

I don't think Typescript will analyze deps vs devDeps anytime soon. So, an alternative is that Typescript could add an option to ignore certain `@types`, ideally through tsconfig.json, since AFAIK there are a lot of projects that splits tsconfig.json for different running environments.

cc thw0rted

**Comment by ImRodry** [comment]
The team at TS said they would address this issue in v5.3 so I don’t think there’s any need to continue chatting about it

**Comment by thw0rted** [comment]
Appreciate the ping, but I'm not actually a maintainer here, I just happened to put together a PR last year that kind of sits adjacent to this problem.

If somebody wants to put together a PR that adds `undici` as a dep for `@types/node`, then re-exports the relevant types into the global namespace, you could probably get a maintainer to review that.  Alternately, you might be able to copy out a "snapshot" of typings from the undici package by hand, with some comments how to keep them up to date -- that would still be a lot better than the current state of affairs, I think.  Either way, I'd be happy to see this closed out.

----

As I was typing the above, @ImRodry said this is supposed to be addressed by TS directly; if anybody has more info about that I'd love to read it.

**Comment by ImRodry** [comment]
It’s in the Iteration Plan under “Community Tooling”. The first point links directly to this issue: https://github.com/microsoft/TypeScript/issues/55486

**Comment by Andrewmat** [comment]
@thw0rted Sorry for the misunderstanding... I mentioned you mostly because you raised some issues about the undici as dep method.

@ImRodry That's where I'm coming from. I saw that the discussion opened up for these conflicts between DOM and Node typings, but it was considered as off topic. My point is that is best way to tackle it. Since typescript maintainers are looking into this issue, it would be good for the discussion to consider that. As it stands now, `setTimeout`, `Blob`, `Buffer` are already inside this scope, and `fetch` is joining them.

**Comment by andrewbranch** [comment]
It was discussed some before, but I want to reiterate for clarity that conflicts between `@types/node` and `lib.dom.d.ts` do need to be avoided at all costs. While it’s not good for a frontend project to end up with `@types/node` in their compilation, it does happen, and it’s not always due to a misconfiguration on the user’s part; often it’s a deep transitive reference stemming from one of their legitimate frontend dependencies. Ideas around having the TypeScript compiler simply refuse to load `@types/node` via one of these transitive references have been pretty thoroughly explored, there’s no proposal that’s anywhere near ready to be adopted.

I tried for an hour or so to use @thw0rted’s conditional type trick (#59905) to import `undici` types and put them into globals in a way that doesn’t conflict with DOM types, but there are two problems:

1. It works well for declaring the value side of `Request`, `Response`, `Headers`, and `FormData`, but the type sides of `Request`, `Response`, and `FormData` don’t merge cleanly with the DOM interfaces. I think the conditional type trick can be used deeper within those types to fix the merge, but the location of the changes would be in the `undici` types themselves.
2. The same conditional type trick doesn’t work for declaring `fetch` itself, because lib.dom.d.ts declares it as `declare function fetch(...)` whereas the conditional type would require us to declare it as `declare var fetch: ...`, and those are not allowed to merge. There are a couple potential options here.
   - `@types/node` could use `declare function fetch(...)` as well, but this would add an overload instead of changing the type based on inclusion of DOM types. Depending on the overload order, this would be non-ideal either for actual Node.js projects or for frontend projects who accidentally have `@types/node` included—someone is going to get the wrong overload. But with `Request` and `Response` being properly substituted, this might not be a huge deal. (This is the current situation with `setTimeout`.)
   - lib.dom.d.ts could switch from `declare function fetch(...)` to `declare var fetch: ...`. The potential downsides to this are that users who are currently declaring a global `fetch` function (perhaps unnecessarily, or maybe they’re intentionally adding a weird custom overload) will start seeing errors, and that users of older TypeScript versions and newer `@types/node` versions would have an error. It also just looks out of place with all the other declared functions 🤷.
   - We could investigate a change in TypeScript itself that allows ambient function-typed variables (what we’d like to put in @types/node) to merge as overloads with function declarations (what’s already in lib.dom.d.ts).

I think problem (2) won’t be a total blocker; it seems likely that we could implement a less-than-perfect solution in the short term and then figure out a way to iterate on it, but some testing will be needed.

For problem (1), I think the best course of action is, somewhat unfortunately, to start copying the `undici` types into DT and making the necessary changes to prevent compilation errors when lib.dom.d.ts is included. I think the changes needed would be too fragile and non-trivial to PR to undici and leave to those maintainers. Small changes could break compatibility with TypeScript in unexpected ways, so I think we want to keep those kinds of hacks in DefinitelyTyped, where TypeScript compatibility is checked every night and TS maintainers can contribute more easily. On the upside, `undici` is fairly large, so copying the types instead of taking a dependency saves a lot of JS weight. Perhaps once the work is done, `undici` will opt to replace its own types with a dependency on `@types/node`, but I don’t think it particularly matters.

**Comment by thw0rted** [comment]
This probably isn't the place to discuss it, but my first thought on reading your opening paragraph was, has the TS team considered taking on a "lib.node.d.ts" so they can be in charge of ensuring conflicts don't happen? It seems precarious to put so much responsibility in the hands of a community maintained library. As you point out, it's probably a rare frontend project that doesn't wind up with `@types/node` installed eventually.

**Comment by pauldraper** [comment]
@thw0rted if anything, the team is moving away from the bundled type declarations.

https://www.npmjs.com/package/@types/web

Though that is a somewhat different question than maintainence.

**Comment by andrewbranch** [comment]
@thw0rted I think both the collaboration between the TS team and the community on `@types/node` and the infrastructure overlap between TS and DT are working well (age of this issue notwithstanding), and there would be little to gain by moving the work into the TypeScript repo. I think we would lose a lot, actually.

**Comment by andrewbranch** [comment]
I think I have both a version that copies undici types and a version that has undici as a dependency working. The latter is satisfyingly short:

```ts
declare namespace NodeJS {
  namespace undici {
    type Request = typeof globalThis extends { onmessage: any } ? {} : import("undici").Request;
    type Response = typeof globalThis extends { onmessage: any } ? {} : import("undici").Response;
    type File = typeof globalThis extends { onmessage: any } ? {} : import("undici").File;
    type FormData = typeof globalThis extends { onmessage: any } ? {} : import("undici").FormData;
    type Headers = typeof globalThis extends { onmessage: any } ? {} : import("undici").Headers;
  }
}

declare function fetch (
  input: import("undici").RequestInfo,
  init?: import("undici").RequestInit
): Promise<Response>;

interface Request extends NodeJS.undici.Request {}
declare var Request: typeof globalThis extends {
  onmessage: any;
  Request: infer T;
} ? T : typeof import("undici").Request;

interface Response extends NodeJS.undici.Response {}
declare var Response: typeof globalThis extends {
  onmessage: any;
  Response: infer T;
} ? T : typeof import("undici").Response;

declare var File: typeof globalThis extends {
  onmessage: any;
  File: infer T;
} ? T : typeof import("undici").File;
interface File extends NodeJS.undici.File {}

declare var FormData: typeof globalThis extends {
  onmessage: any;
  FormData: infer T;
} ? T : typeof import("undici").FormData;
interface FormData extends NodeJS.undici.FormData {}

declare var Headers: typeof globalThis extends {
  onmessage: any;
  Headers: infer T;
} ? T : typeof import("undici").Headers;
interface Headers extends NodeJS.undici.Headers {}
```

> It works well for declaring the value side of `Request`, `Response`, `Headers`, and `FormData`, but the type sides of `Request`, `Response`, and `FormData` don’t merge cleanly with the DOM interfaces. I think the conditional type trick can be used deeper within those types to fix the merge, but the location of the changes would be in the undici types themselves.

Upon taking a closer look at how @thw0rted made the `Blob` global work, I realized I had missed something, so it does appear to be possible to use `undici` unchanged if that’s the route we want to go.

> The same conditional type trick doesn’t work for declaring `fetch` itself... `@types/node` could... add an overload instead of changing the type... Depending on the overload order, this would be non-ideal either for actual Node.js projects or for frontend projects who accidentally have `@types/node` included—someone is going to get the wrong overload. But with `Request` and `Response` being properly substituted, this might not be a huge deal.

Yeah. The signatures are identical with the other globals substituted, so adding the overload works perfectly fine.

Planning to chat with the rest of the team tomorrow about which approach to PR.

**Comment by andrewbranch** [comment]
The feedback I got from the team was that we should try to avoid the undici-as-dependency approach mostly due to the significant impact it would have on `@types/node` total installation size. It was suggested that we could make a `@types/undici` rather than copying them into `@types/node` itself, which would make it easier for `undici` itself to drop their included types if they’re interested in doing that.

**Comment by Qix-** [comment]
Well DT is good for packages that don't have types already but for packages that do, or are otherwise written in TS, DT is usually the worse of the two options due to desynchronization. If there could be a way to extract types at build time, prior to publishing, that'd be the 'ideal' way depending on how you look at things.

**Comment by piranna** [comment]
Prepublish script?

**Comment by jchitel** [comment]
It seems to me that this points to a fundamental problem with how TypeScript handles type definitions, particularly with global types. The actual TypeScript team should be more involved in this discussion.

I think the best solution today is to use a clever combination of the `lib` and `types` settings within *multiple* tsconfig files in a single project.

An example web project might have 3 "environments" that are used by various files: a "web" environment for production code that will run in a browser, a "test" environment that runs in Node (or Deno or Bun) and polyfills web APIs using something like `happy-dom` or `jsdom`, and a "server" environment that includes everything running in a regular Node (or Deno or Bun) environment with no web APIs.

This project can handle the differences in available global APIs and modules by defining 3 separate tsconfig files with explicitly defined `lib` and `types` settings:
- `tsconfig.web.json`:
  - `"lib": ["dom", "dom.iterable", "..."]`
  - `"types": []` (explicitly declare no types to be included)
  - `"include": ["./client"]`
  - `"exclude": ["./client/**/__tests__"]` (exclude tests, which are covered below)
- `tsconfig.node.json`:
  - `"lib": ["..."]` (no DOM APIs are available)
  - `"types": ["node"]` (explicitly include only Node types)
  - `"include": ["./server", "..."]` (add whatever other files run in Node, like `webpack.config.ts` or `vite.config.ts`)
- `tsconfig.test.json`:
  - `"lib": ["dom", "dom.iterable", "..."]`
  - `"types": ["node"]` (explicitly include only Node types)
  - `"include": ["./client/**/__tests__"]` (only include tests)

This solves most of the problem, notably for files that run only on the web or only in Node. However, the "test" environment has both DOM and Node types. Technically we can't know which `fetch` types are correct, since any DOM polyfill that is used may or may not overwrite Node's built-in `fetch` implementation.

TypeScript's solution for this problem up to this point has been to merge both definitions into one. This works well for interfaces, which will simply accept properties from both definitions. It does not work well for functions, however. We already see this problem with `setTimeout`. The DOM types return a `number`, while the Node types return a `NodeJS.Timeout`. The simple workaround on the web side is to use `window.setTimeout`, which resolves to only the DOM version. But there is no good solution on the Node side. In the case of `setTimeout` things typically just happen to work, but if there was a more complicated difference between the types, this would have been a problem ages ago.

As more and more web APIs are added to non-browser environments, this problem is likely to only get worse. For example, since `fetch` was designed to work in the browser sandbox, it supports only a subset of HTTP (e.g. no direct access to all cookies), so some environments add additional properties to the options object passed to `fetch`. Other web APIs are likely to have similar situations.

We need the ability to more finely tune what an environment looks like, down to selecting individual items from various type definitions. This might be possible using conditional types and `Pick` and `Omit`, but those don't really feel "first-class" IMO. I'd like to hear what other people's thoughts are on this.

**Comment by andrewbranch** [comment]
People’s thoughts on this:

- https://github.com/microsoft/TypeScript/issues/31894
- https://github.com/microsoft/TypeScript/issues/19183
- https://github.com/microsoft/TypeScript/issues/9466
- https://github.com/microsoft/TypeScript/issues/3538
- https://github.com/microsoft/TypeScript/issues/36057
- https://github.com/microsoft/TypeScript/issues/52433
- https://github.com/microsoft/TypeScript/issues/43972

> The actual TypeScript team should be more involved in this discussion.

😕

**Comment by jchitel** [comment]
> > The actual TypeScript team should be more involved in this discussion.
> 
> 😕

Oops! I didn't notice that someone from the TypeScript team was already here! I read the earlier comments more thoroughly than the latter ones, so I completely missed your responses. I'm sorry if I offended anyone.

**Comment by andrewbranch** [comment]
No worries 😄 I think you’ll find some of those links interesting. I think this issue is probably not the best place to go deep into the weeds on rethinking environmental types from the ground up. I broadly agree with you, and this problem used to be very much top-of-mind for me. But from my perspective, each time we’ve hit a wave of global type clashes that seems really intractable, we’ve actually been able to resolve things _pretty_ well with the tools we already have. The PR I have up to add `fetch` typings, closing this issue, is an example. It’s not perfect but it’s _really_ close, I think. Another one:

> In the case of `setTimeout` things typically just happen to work

Again, it’s not perfect, but it doesn’t provide a super strong justification for overhauling the way global types work. We just try our best, through DT contributions, to discourage Node.js-related typings from pulling in the DOM unnecessarily and vice versa, so users can _hopefully_ avoid having `@types/node` and `lib.dom.d.ts` in the same compilation, while ensuring they’re written in a way that doesn’t explode into errors when they _are_ in the same compilation. And this continues to mostly work out for most cases.

I would love to have a more rigorous solution in the future, but it’s definitely not worth holding off on adding `fetch` while we wait for a silver bullet.

**Comment by jchitel** [comment]
That makes a lot of sense. The Issues you posted are super interesting and show that others have clearly thought more about this than I have 😅

I agree that there needs to be a solution right now to resolve this particular issue. I'll defer further discussion to your PR.

**Comment by andrewbranch** [comment]
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66824 is ready. It only updates v20 typings due to a test infrastructure limitation that will likely be resolved within a couple weeks. I’d like to wait for a non-TS team `@types/node` maintainer to approve.

**Comment by bootrino** [comment]
This is a giant thread and all i want is to know how to use the correct types for native fetch in nodejs.

**Comment by Jokinen** [comment]
Apologies for replying to such an old thread, but I am not sure what could be a more suitable place.

In the introduction of these types there was a workaround that allowed the types to roughly correct for pure node applications and applications with code that should be ran in the browser as well. This combination seems to have become an obstacle in my case.

I think I am working with a relatively common setup where I have a TS project with code that runs on node and code that runs int he browser (in this case a next application). With such an application `lib.dom.d.ts` is likely included in the tsconfig, but even if was not, it's not unlikely for it to be pulled in by some dependency. In essence, for this kind of a project it's difficult to get rid of `lib.dom.d.ts`.

The presence of `lib.dom.d.ts` means that the used types are the DOM types. As the original PR states, these are very similar with undici's types. However, there are some difference with one being the omission of the `dispatcher` field. This field can be use to set a custom agent, which allows custom keepalive settings to be configured.

I've used a better part of the day trying to find the best workaround, but it seems that there are quite a few potholes that can be struck. For instance, importing `fetch` directly from `undici` breaks interop with msw.

Ultimately I was able to work around the limitation by using two `tsconfig` files and tinkering with the settings and project structure until I was able isolate our node fetch integration from `lib.dom.d.ts`.

**Comment by thw0rted** [comment]
Do you have an example of the code that wasn't working in the mixed environment (Node code, with lib-dom included)?  I'm wondering if some selective type-casting could have saved you the trouble of multiple tsconfig files.  That's probably a good long-term solution, though.

**Comment by Jokinen** [comment]
@thw0rted Thank you for the reply. I ended up iterating and did end up going the route you suggested. I ended up installing `undici-types` and type cast in a few places in order to get the types to cooperate. We have a wrapper on top of the `fetch` API which made the process easier.

All global types were browser types so whenever we attempted to use node-only-features, we'd have issues. In our case the problem case was adding a `dispatcher` that changed the default keepalive settings.

```ts
fetch('https://example.com', {
  // ...options
  dispatcher: const customKeepAliveAgent = new undici.Agent({
    // customs settings
  });,
});
```

I can also vouch that it doesn't feel like a good long-term solution. As we've implemented more functionality to our `fetch` wrapper, we've come across type issues consistently. In time we'll likely find a better balance in casting and we'll have fewer issues.

The type inconsistencies can be difficult to reason about if you aren't aware of the limitations of the type system. For instance, we ended up breaking `instanceof URL` checks. I assume it's due to a type cast that's not optimal. Likely we've cast a `RequestInfo` to be a `Undici.RequestInfo` (e.g. `string | node:url.URL | Undici.Request`. When we make the comparison TS sees the `URL` in `instanceof URL` to be a `browser.URL`. It basically results in a scenario like this:

```ts
export const getUrl = (url: Undici.RequestInfo) => {
  if (typeof url === 'string') {
    return url;
  } else if (url instanceof URL) {
    return url.href;
  }

  // @ts-expect-error: TS errors because field 'url' does not exist in 'URL'
  return url.url;
};
```
